### PR TITLE
Feature/bt133 add contract

### LINKED ISSUE #31: Add smart contract support to the CENNZnut spec
Add a contracts section to the CENNZnut format in order to support smart contract permissioning.
It will be a whitelist of authorized contracts. It signals the runtime should allow a contract to be called given it's address is included.

It will need to be open for extension to handle whitelisting methods and parameters on contracts also.

```js
contracts:
    "5DZYAPfcRGtU7J9xLL8nfC76Hszts5nYsYExpQ4kQeUEPRbF": {},
    "5EAH7hnTzpDxx9X7Kx1476LEwax1Uc4oGy7tATKar45TfNsG": {
        // future extension: `*` whitelist is allow all
        "methods": "*"
    },
    "5FjDaDVeuhSdGFWzyvfd56aZnhUrK3pVKMTApyYAeCw2woSe": {
        // future extension:  allow whitelisted methods only
        "methods": {
            "do the thing": {}
        }
    }
},
```

- [ ] Update the CENNZnut spec.
- [ ] Implementation

### PR DESCRIPTION
**Pull request by hoani**
* Add contract module
* Module count range is now 0 to 255 (previously 1 to 256)
* Wildcard contracts have address [0x00; 32]

Hoani concerns:
* Not too happy with the amount of times `[u8; 32]` shows up in the code
* The `Validate` interface may not need `validate` anymore
* Testing could be split a little nicer between modules (`contracts` is a good example of how this could work), but I don't think this PR is the time to do it.
* the names `module`, `modules` and `Module` could potentially be better named `runtime`, but the doughnut spec specifies `module`

Closes #31

### CODE REVIEW

**Comment by alexsednz** [comment]
Just a suggestion for the future. The branch name would better start with the public issue number. Here 31.

File: src/cennznut/mod.rs [hunk 1]
```diff
@@ -87,15 +101,35 @@ impl Decode for CENNZnut {
...
+        self.validate_runtime_call(module_name, method_name, args)
+    }
+
+    fn validate_runtime_call(
+        &self,
+        module_name: &str,
+        method_name: &str,
+        args: &[PactType],
+    ) -> Result<(), ValidationErr<RuntimeDomain>> {
+        match &self {
```
Comment (alexsednz) [inline src/cennznut/mod.rs]: validate_module, validate_contract and so on should belong to CENNZnut. We shouldn't check for the version everywhere.

File: src/validation.rs [hunk 2]
```diff
@@ -37,11 +37,40 @@ impl<Domain: Display> Display for ValidationErr<Domain> {
     }
 }
 
-pub trait Validate<Domain: Display> {
+pub trait Validate<RuntimeDomain: Display, ContractDomain: Display> {
```
Comment (alexsednz) [inline src/validation.rs]: Looks like Validate could be decomposed into two traits:
```
pub trait ValidateRuntimeCall {
fn validate(..
}

pub trait ValidateContractCall {
fn validate(..
}
```
This way, the separation of the returned error type could make more sense and also there could be no redundant validate function.

File: src/cennznut/v0/contract.rs [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+
+/// A CENNZnet permission domain contract
+#[cfg_attr(test, derive(Clone, Debug, Eq, PartialEq))]
+pub struct Contract {
+    pub address: [u8; 32],
+    pub block_cooldown: Option<u32>,
+}
+
+impl Contract {
+    pub fn new(address: &[u8; 32]) -> Self {
```
Comment (alexsednz) [inline src/cennznut/v0/contract.rs]: Doc

File: src/cennznut/v0/contract.rs [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+        Self {
+            address: *address,
+            block_cooldown: None,
+        }
+    }
+
+    pub fn wildcard() -> Self {
```
Comment (alexsednz) [inline src/cennznut/v0/contract.rs]: Doc

File: src/cennznut/v0/contract.rs [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        Self {
+            address: [0x00; 32],
+            block_cooldown: None,
+        }
+    }
+
+    pub fn block_cooldown(mut self, block_cooldown: u32) -> Self {
```
Comment (alexsednz) [inline src/cennznut/v0/contract.rs]: Doc

File: src/cennznut/mod.rs (hunk 1 above)
Comment (hoani) [reply src/cennznut/mod.rs]: Unsure what is meant by this?

It is possible that we could move `validate` to be part of the versioned CennzNut, but this PR doesn't address that design decision

File: src/cennznut/v0/contract.rs [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    }
+}
+
+impl Encode for Contract {
+    fn encode_to<T: Output>(&self, buf: &mut T) {
+        let mut has_cooldown_byte = 0x00_u8;
+        if self.block_cooldown.is_some() {
+            has_cooldown_byte |= 0b0000_0001_u8;
+        }
+        buf.push_byte(has_cooldown_byte.swap_bits());
```
Comment (alexsednz) [inline src/cennznut/v0/contract.rs]: if we had done ```has_cooldown_byte |= 0b1000_0000_u8;``` would we still need to do swap_bits?

File: src/cennznut/v0/mod.rs [hunk 7]
```diff
@@ -99,30 +133,139 @@ impl Decode for CENNZnutV0 {
     }
 }
 
-impl Validate<ModuleDomain> for CENNZnutV0 {
-    /// Validates a CENNZnut by (1) looking for `module_name` and `method_name` and (2) executing the
-    /// Pact interpreter if constraints exist
-    fn validate(
+impl CENNZnutV0 {
```
Comment (alexsednz) [inline src/cennznut/v0/mod.rs]: We could implement the two traits ValidateContractCall and ValidateRuntimeCall that I mentioned in my other comment for both CENNZnutV0 and CENNZnut. In the latter case, CENNZnut would check the version and call the same interface from the corresponding version of CENNZnutV0.

File: src/cennznut/v0/contract.rs [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+    }
+
+    pub fn block_cooldown(mut self, block_cooldown: u32) -> Self {
+        self.block_cooldown = Some(block_cooldown);
+        self
+    }
+}
+
+impl Encode for Contract {
+    fn encode_to<T: Output>(&self, buf: &mut T) {
```
Comment (alexsednz) [inline src/cennznut/v0/contract.rs]: Have we remained backwards compatible after implementing encode and decode here?

**Comment by alexsednz** [comment]
Thank you @hoani especially for adding a very generous amount of tests alongside your changes. I just left a few comments for you to consider.

File: src/cennznut/v0/contract.rs [hunk 9]
```diff
@@ -0,0 +1,293 @@
+// Copyright (C) 2019-2020 Centrality Investments Limited
+//!
+//! # CENNZnut - Contracts
+//!
+//! Delegated smart contract permissioning of CENNZnut for use in CENNZnet
+//!
+
+#![warn(clippy::pedantic)]
+
+#[cfg(not(feature = "std"))]
```
Comment (holygits) [inline src/cennznut/v0/contract.rs]: This should be done in one top level place e.g. `lib.rs` and `alloc` just imported here

File: src/cennznut/v0/contract.rs [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+#[cfg(feature = "std")]
+extern crate std as alloc;
+
+use bit_reverse::ParallelReverse;
+use codec::{Decode, Encode, Input, Output};
+
+/// A CENNZnet permission domain contract
+#[cfg_attr(test, derive(Clone, Debug, Eq, PartialEq))]
+pub struct Contract {
+    pub address: [u8; 32],
```
Comment (holygits) [inline src/cennznut/v0/contract.rs]: You could alias `[u8; 32]` as `type ContractAddress = [u8; 32];`

File: src/cennznut/v0/contract.rs [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+    pub fn new(address: &[u8; 32]) -> Self {
+        Self {
+            address: *address,
+            block_cooldown: None,
+        }
+    }
+
+    pub fn wildcard() -> Self {
+        Self {
+            address: [0x00; 32],
```
Comment (holygits) [inline src/cennznut/v0/contract.rs]: Might seem silly but it's worth using a constant for this

File: src/test.rs [hunk 12]
```diff
@@ -14,9 +14,9 @@ use std::vec::Vec;
 #[test]
 fn it_fails_decode_with_invalid_constraints() {
     let encoded_cennznut: Vec<u8> = vec![
-        0, 0, 0, 64, 109, 111, 100, 117, 108, 101, 95, 116, 101, 115, 116, 0, 0, 0, 0, 0, 0, 0, 0,
-        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 109, 101, 116, 104, 111, 100, 95, 116, 101, 115,
-        116, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0x80, 64, 109, 111, 100, 117, 108, 101, 95, 116, 101, 115, 116, 0, 0, 0, 0, 0, 0, 0,
```
Comment (holygits) [inline src/test.rs]: Would prefer if we didn't mix hex values in here

File: src/cennznut/v0/contract.rs (hunk 11 above)
Comment (hoani) [reply src/cennznut/v0/contract.rs]: Not silly, this really bugged me to tbh

File: src/cennznut/v0/contract.rs (hunk 10 above)
Comment (hoani) [reply src/cennznut/v0/contract.rs]: Nice!

File: src/cennznut/v0/contract.rs (hunk 9 above)
Comment (hoani) [reply src/cennznut/v0/contract.rs]: Done

File: src/test.rs (hunk 12 above)
Comment (hoani) [reply src/test.rs]: Done

**Comment by alexsednz** [comment]
Very nice improvements. To me, the only important remaining place for improvement is swap_bits. If you can find a way to avoid them, it may simplify/streamline your solution.

**Comment by hoani** [comment]
Thanks Alex, unfortunately the swap bits are required because the cennznut protocol requires little endian encoding. The swap_bits design wasnt implemented in this PR, so if we wanted to change it, it would have to be a seperate PR

File: src/cennznut/mod.rs [hunk 13]
```diff
@@ -87,15 +96,77 @@ impl Decode for CENNZnut {
...
+
+        assert_eq!(
+            cennznut.validate_runtime_call(&module.name, &method.name, &[]),
+            Ok(())
+        );
+    }
+
+    #[test]
+    fn it_validates_v0_contracts() {
+        let modules: Vec<(String, Module)> = Vec::default();
```
Comment (holygits) [inline src/cennznut/mod.rs]: `Vec<(ContractAddress, Contract)>::default()` and same in other places

File: src/cennznut/v0/contract.rs [hunk 14]
```diff
@@ -0,0 +1,287 @@
+// Copyright (C) 2019-2020 Centrality Investments Limited
+//!
+//! # CENNZnut - Contracts
+//!
+//! Delegated smart contract permissioning of CENNZnut for use in CENNZnet
+//!
+
+use crate::cennznut::ContractAddress;
```
Comment (holygits) [inline src/cennznut/v0/contract.rs]: `crate::cennznut::{ContractAddress, CONTRACT_WILDCARD};`

File: src/cennznut/v0/contract.rs [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+
+    pub fn block_cooldown(mut self, block_cooldown: u32) -> Self {
+        self.block_cooldown = Some(block_cooldown);
+        self
+    }
+}
+
+impl Encode for Contract {
+    fn encode_to<T: Output>(&self, buf: &mut T) {
+        let mut has_cooldown_byte = 0x00_u8;
```
Comment (holygits) [inline src/cennznut/v0/contract.rs]: No `mut` and maybe more clear
```
let has_cooldown_byte = if self.block_cooldown.is_some() {
  0b0000_0001;
} else {
  0;
};
```

File: src/cennznut/v0/contract.rs [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+        }
+    }
+}
+
+impl Decode for Contract {
+    fn decode<I: Input>(input: &mut I) -> Result<Self, codec::Error> {
+        let has_cooldown_byte: u8 = input.read_byte()?.swap_bits();
+        let has_cooldown: bool = (has_cooldown_byte & 0x01) == 0x01;
+
+        let mut address: ContractAddress = Default::default();
```
Comment (holygits) [inline src/cennznut/v0/contract.rs]: `ContractAddress::default()` for consistency

File: src/cennznut/v0/mod.rs [hunk 17]
```diff
@@ -24,21 +15,27 @@ use bit_reverse::ParallelReverse;
...
+use super::{ContractDomain, RuntimeDomain};
 use crate::PartialDecode;
-use crate::Validate;
 use crate::ValidationErr;
+
+use contract::Contract;
 use module::Module;
 
-pub const WILDCARD: &str = "*";
+use super::WILDCARD;
```
Comment (holygits) [inline src/cennznut/v0/mod.rs]: Can we compress this imports in the `::{<a>, <b>, <c>}` style

File: src/cennznut/mod.rs [hunk 18]
```diff
@@ -87,15 +96,77 @@ impl Decode for CENNZnut {
     }
 }
 
-impl Validate<ModuleDomain> for CENNZnut {
-    fn validate(
+impl CENNZnut {
+    pub fn validate_runtime_call(
```
Comment (holygits) [inline src/cennznut/mod.rs]: Should add some doc comments for these

File: src/cennznut/v0/contract.rs [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+        assert_eq!(contract.address[18], expected_18);
+        assert_eq!(contract.address[27], expected_27);
+    }
+
+    #[test]
+    fn it_initializes_wildcard() {
+        let expected_length: usize = 32;
+        let expected_value: u8 = 0x00;
+
+        let contract = Contract::wildcard();
```
Comment (holygits) [inline src/cennznut/v0/contract.rs]: 👍

File: src/cennznut/v0/contract.rs [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+
+        assert_eq!(contract.address.len(), expected_length);
+        for i in 0..32 {
```
Comment (holygits) [inline src/cennznut/v0/contract.rs]: Should use wildcard address const for this test

**Pull request merged**